ignore image err when docker image has been cleaned up

xiaoanyunfei opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Kubelet gets stuck trying to inspect a container whose image has been cleaned up

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #https://github.com/kubernetes/kubernetes/issues/84214

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/sig node

xiaoanyunfei commented:
/assign @yujuhong

mattjmcnaughton commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -347,7 +347,7 @@ func (ds *dockerService) ContainerStatus(_ context.Context, req *runtimeapi.Cont
 	ir, err := ds.client.InspectImageByID(r.Image)
 	if err != nil {
-		return nil, fmt.Errorf("unable to inspect docker image %q while inspecting docker container %q: %v", r.Image, containerID, err)
+		klog.Errorf("unable to inspect docker image %q while inspecting docker container %q: %v", r.Image, containerID, err)
```
I'm a little nervous about ignoring this error. I see how ignoring it could be useful when the image has already been cleaned up, but are there other times when we would want failures to raise an error?

zouyee replied:
why  ignore it nor return err？

mattjmcnaughton commented on pkg/kubelet/dockershim/convert.go:
```diff
@@ -69,7 +69,7 @@ func toPullableImageID(id string, image *dockertypes.ImageInspect) string {
 	// the first digest instead.
 	imageID := DockerImageIDPrefix + id
-	if len(image.RepoDigests) > 0 {
+	if image != nil && len(image.RepoDigests) > 0 {
```
Can you share more about why this nil check is necessary?

xiaoanyunfei replied:
We can call `toPullableImageID` to generate default imageID when `InspectImageByID` return `nil, err`

mattjmcnaughton commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -347,7 +347,7 @@ func (ds *dockerService) ContainerStatus(_ context.Context, req *runtimeapi.Cont
 	// Convert the image id to a pullable id.
 	ir, err := ds.client.InspectImageByID(r.Image)
```
Could one possible option be checking that an image with the `id` exists? If not, we could return without error.

xiaoanyunfei replied:
But the other `InspectImageByID` err  can also case kubelet stuch

mattjmcnaughton reviewed (commented):
Thanks for looking at this issue report @xiaoanyunfei !

@zouyee mind weighing in here? You mentioned you were unable to reproduce the original error.

xiaoanyunfei commented:
/test pull-kubernetes-integration

mattjmcnaughton commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -347,7 +347,10 @@ func (ds *dockerService) ContainerStatus(_ context.Context, req *runtimeapi.Cont
+		if !libdocker.IsImageNotFoundError(err) {
+			return nil, fmt.Errorf("unable to inspect docker image %q while inspecting docker container %q: %v", r.Image, containerID, err)
+		}
+		klog.Warningf("ignore error image %q not found  while inspecting docker container %q: %v", r.Image, containerID, err)
```
Nit: There are two spaces between `found` and `while`.

mattjmcnaughton reviewed (approved):
/lgtm

I'm comfortable with this change after giving it some thought. Even if this error isn't a common occurrence, it makes sense to me that we'd want to gracefully handle it by outputting the default image id, instead of the whole ContainerStatus call returning an error. 

Definitely curious to hear from others in the community thought :)

Commit:
ignore image err when docker image has been cleaned up

xiaoanyunfei commented:
> Thanks for looking at this issue report @xiaoanyunfei !
> 
> @zouyee mind weighing in here? You mentioned you were unable to reproduce the original error.

I reproduced once on `master` branch, but never again.

I updated the pod image and deleted the original image, 
then deleted the pod, and it got stuck. so I restarted kubelet, the image error occured and the pod was termianting

mattjmcnaughton reviewed (approved):
/lgtm

Re-adding lgtm after commit to fix typo.

dims commented:
/uncc

yujuhong commented:
/lgtm
/approve

xiaoanyunfei commented:
/test pull-kubernetes-integration
